redfish-core/lib/power_supply_metrics.hpp

```diff
@@ -187,110 +189,199 @@ inline void
     // It will have 0 to many date/timestamp, average, and maximum entries.
     aResp->res
         .jsonValue["Oem"]["IBM"]["InputPowerHistoryItems"]["@odata.type"] =
         "#OemPowerSupplyMetric.InputPowerHistoryItems";
 
-    const std::array<const char*, 2> interfaces = {
-        "org.open_power.Sensor.Aggregation.History.Average",
-        "org.open_power.Sensor.Aggregation.History.Maximum"};
+    std::string serviceName;
+    std::string averagePath;
+    std::string maximumPath;
 
-    crow::connections::systemBus->async_method_call(
-        [aResp, chassisID, powerSupplyID](
+    getServicePathValues(aResp, inputHistoryItem[0], inputHistoryItem[1],
+                         serviceName, averagePath, maximumPath);
+
+    BMCWEB_LOG_DEBUG << "EXIT: getValues(...)";
+}
+
+/**
+ * @brief Retrieves valid input history item.
+ *
+ * Not all power supplies support the power input history. Do not provide
+ * Redfish fields for input power history if no associated endpoint matches this
+ * chassis.
+ *
+ * @param asyncResp     Pointer to object holding response data
+ * @param powerSupplyPath Validated power supply path
+ * @param callback      Callback for next step to get valid chassis ID
+ */
+template <typename Callback>
+inline void
+    getValidInputHistory(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+                         const std::string& powerSupplyPath,
+                         Callback&& callback)
+{
+    BMCWEB_LOG_DEBUG << "getValidInputHistory enter";
+    BMCWEB_LOG_DEBUG << "powerSupplyPath: " << powerSupplyPath;
+
+    auto respHandler =
+        [callback{std::move(callback)}, asyncResp, powerSupplyPath](
             const boost::system::error_code ec,
-            const std::vector<std::pair<
-                std::string,
-                std::vector<std::pair<std::string, std::vector<std::string>>>>>&
-                intfSubTree) mutable {
+            const std::variant<std::vector<std::string>>& endpoints) {
+            BMCWEB_LOG_DEBUG << "getValidInputHistory respHandler enter";
+
             if (ec)
             {
-                BMCWEB_LOG_DEBUG << "D-Bus response error on GetSubTree " << ec;
-                messages::internalError(aResp->res);
+                BMCWEB_LOG_ERROR
+                    << "getValidInputHistory respHandler D-Bus error: " << ec;
+                messages::internalError(asyncResp->res);
                 return;
             }
 
-            std::string serviceName;
-            std::string averagePath;
-            std::string maximumPath;
+            // Set the default value to resourceNotFound, and if we confirm
+            // finding association with chassisID and powerSupplyID is correct,
+            // the error response will be cleared.
+            messages::resourceNotFound(asyncResp->res, "PowerSupplyMetrics",
+                                       "Metrics");
+
+            const std::vector<std::string>* inputHistoryItem =
```

Comment:
Nit: Should this be inputHistoryItems (with an 's'), since there are multiple items?

Reply:
See prior explanation.

I will treat it like sheep. One sheep, and then is it two sheep or two sheeps? :-)